File: crypto/crypto_test.go
// Copyright 2026 Marcelo Cantos
// SPDX-License-Identifier: Apache-2.0

package crypto

import (
	"bytes"
	"crypto/ecdh"
	"crypto/rand"
	"fmt"
	"testing"
)

func TestKeyExchangeAndEncrypt(t *testing.T) {
	// Simulate client and server generating key pairs.
	client, err := GenerateKeyPair()
	if err != nil {
		t.Fatal(err)
	}
	server, err := GenerateKeyPair()
	if err != nil {
		t.Fatal(err)
	}

	// Derive session keys — client sends, server receives.
	clientSendKey, err := DeriveSessionKey(client.Private, server.Public, []byte("client-to-server"))
	if err != nil {
		t.Fatal(err)
	}
	serverRecvKey, err := DeriveSessionKey(server.Private, client.Public, []byte("client-to-server"))
	if err != nil {
		t.Fatal(err)
	}

	// Keys should match.
	if !bytes.Equal(clientSendKey, serverRecvKey) {
		t.Fatal("derived keys don't match")
	}

	// Reverse direction.
	clientRecvKey, err := DeriveSessionKey(client.Private, server.Public, []byte("server-to-client"))
	if err != nil {
		t.Fatal(err)
	}
	serverSendKey, err := DeriveSessionKey(server.Private, client.Public, []byte("server-to-client"))
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(clientRecvKey, serverSendKey) {
		t.Fatal("reverse derived keys don't match")
	}

	// Create channels.
	clientCh, err := NewChannel(clientSendKey, clientRecvKey)
	if err != nil {
		t.Fatal(err)
	}
	serverCh, err := NewChannel(serverSendKey, serverRecvKey)
	if err != nil {
		t.Fatal(err)
	}

	// Client → server.
	msg := []byte("hello from client")
	ct := clientCh.Encrypt(msg)
	pt, err := serverCh.Decrypt(ct)
	if err != nil {
		t.Fatal("server decrypt:", err)
	}
	if !bytes.Equal(pt, msg) {
		t.Fatalf("got %q, want %q", pt, msg)
	}

	// Server → client.
	msg2 := []byte("hello from server")
	ct2 := serverCh.Encrypt(msg2)
	pt2, err := clientCh.Decrypt(ct2)
	if err != nil {
		t.Fatal("client decrypt:", err)
	}
	if !bytes.Equal(pt2, msg2) {
		t.Fatalf("got %q, want %q", pt2, msg2)
	}
}

func TestSymmetricChannel(t *testing.T) {
	secret, err := GenerateSecret()
	if err != nil {
		t.Fatal(err)
	}

	clientCh, err := NewSymmetricChannel(secret, false)
	if err != nil {
		t.Fatal(err)
	}
	serverCh, err := NewSymmetricChannel(secret, true)
	if err != nil {
		t.Fatal(err)
	}

	// Client → server.
	msg := []byte(`{"type":"auth","device":"abc123"}`)
	ct := clientCh.Encrypt(msg)
	pt, err := serverCh.Decrypt(ct)
	if err != nil {
		t.Fatal("server decrypt:", err)
	}
	if !bytes.Equal(pt, msg) {
		t.Fatalf("got %q, want %q", pt, msg)
	}

	// Server → client.
	msg2 := []byte(`{"type":"auth_ok"}`)
	ct2 := serverCh.Encrypt(msg2)
	pt2, err := clientCh.Decrypt(ct2)
	if err != nil {
		t.Fatal("client decrypt:", err)
	}
	if !bytes.Equal(pt2, msg2) {
		t.Fatalf("got %q, want %q", pt2, msg2)
	}

	// Multiple messages in sequence.
	for i := 0; i < 100; i++ {
		msg := []byte("message from client")
		ct := clientCh.Encrypt(msg)
		pt, err := serverCh.Decrypt(ct)
		if err != nil {
			t.Fatalf("round %d: %v", i, err)
		}
		if !bytes.Equal(pt, msg) {
			t.Fatalf("round %d: mismatch", i)
		}
	}
}

func TestReplayRejected(t *testing.T) {
	secret, err := GenerateSecret()
	if err != nil {
		t.Fatal(err)
	}
	clientCh, err := NewSymmetricChannel(secret, false)
	if err != nil {
		t.Fatal(err)
	}
	serverCh, err := NewSymmetricChannel(secret, true)
	if err != nil {
		t.Fatal(err)
	}

	ct := clientCh.Encrypt([]byte("first"))
	if _, err := serverCh.Decrypt(ct); err != nil {
		t.Fatal(err)
	}

	// Replay the same ciphertext — should fail.
	if _, err := serverCh.Decrypt(ct); err == nil {
		t.Fatal("expected replay to be rejected")
	}
}

func TestDeriveKeyFromSecret(t *testing.T) {
	secret, err := GenerateSecret()
	if err != nil {
		t.Fatal(err)
	}
	nonce, err := GenerateNonce()
	if err != nil {
		t.Fatal(err)
	}

	key1, err := DeriveKeyFromSecret(secret, nonce)
	if err != nil {
		t.Fatal(err)
	}
	key2, err := DeriveKeyFromSecret(secret, nonce)
	if err != nil {
		t.Fatal(err)
	}

	if !bytes.Equal(key1, key2) {
		t.Fatal("same inputs should produce same key")
	}

	// Different nonce → different key.
	nonce2, _ := GenerateNonce()
	key3, err := DeriveKeyFromSecret(secret, nonce2)
	if err != nil {
		t.Fatal(err)
	}
	if bytes.Equal(key1, key3) {
		t.Fatal("different nonces should produce different keys")
	}
}

func newTestChannel(t *testing.T) (*Channel, *Channel) {
	t.Helper()
	secret, err := GenerateSecret()
	if err != nil {
		t.Fatal(err)
	}
	clientCh, err := NewSymmetricChannel(secret, false)
	if err != nil {
		t.Fatal(err)
	}
	serverCh, err := NewSymmetricChannel(secret, true)
	if err != nil {
		t.Fatal(err)
	}
	return clientCh, serverCh
}

func newTestDatagramChannel(t *testing.T) (*Channel, *Channel) {
	t.Helper()
	secret, err := GenerateSecret()
	if err != nil {
		t.Fatal(err)
	}

	sendKey, err := hkdfDerive(secret, []byte("client-to-server"))
	if err != nil {
		t.Fatal(err)
	}
	recvKey, err := hkdfDerive(secret, []byte("server-to-client"))
	if err != nil {
		t.Fatal(err)
	}

	clientCh, err := NewDatagramChannel(sendKey, recvKey)
	if err != nil {
		t.Fatal(err)
	}
	serverCh, err := NewDatagramChannel(recvKey, sendKey)
	if err != nil {
		t.Fatal(err)
	}
	return clientCh, serverCh
}

func TestDatagramModeAcceptsGaps(t *testing.T) {
	clientCh, serverCh := newTestDatagramChannel(t)

	// Encrypt seq 0, 1, 2, 3, 4, 5 on the sender side.
	cts := make([][]byte, 6)
	for i := range cts {
		cts[i] = clientCh.Encrypt([]byte("msg"))
	}

	// Receiver gets seq 0, 1, and 5 — skipping 2, 3, 4.
	for _, idx := range []int{0, 1, 5} {
		if _, err := serverCh.Decrypt(cts[idx]); err != nil {
			t.Fatalf("seq %d: unexpected error: %v", idx, err)
		}
	}
}

func TestDatagramModeRejectsReplay(t *testing.T) {
	clientCh, serverCh := newTestDatagramChannel(t)

	ct := clientCh.Encrypt([]byte("first"))
	if _, err := serverCh.Decrypt(ct); err != nil {
		t.Fatal(err)
	}
	// Replay the same packet.
	if _, err := serverCh.Decrypt(ct); err == nil {
		t.Fatal("expected replay to be rejected in datagram mode")
	}
}

func TestDatagramModeRejectsOldSeq(t *testing.T) {
	clientCh, serverCh := newTestDatagramChannel(t)

	// Encrypt 6 packets (seq 0..5).
	cts := make([][]byte, 6)
	for i := range cts {
		cts[i] = clientCh.Encrypt([]byte("msg"))
	}

	// Receive seq=5 first (jump forward).
	if _, err := serverCh.Decrypt(cts[5]); err != nil {
		t.Fatal(err)
	}
	// Now seq=3 is in the past — must be rejected.
	if _, err := serverCh.Decrypt(cts[3]); err == nil {
		t.Fatal("expected old sequence number to be rejected in datagram mode")
	}
}

func TestStrictModeRejectsGaps(t *testing.T) {
	clientCh, serverCh := newTestChannel(t)

	// Encrypt two packets.
	ct0 := clientCh.Encrypt([]byte("first"))
	ct1 := clientCh.Encrypt([]byte("second"))
	_ = ct0

	// Receiver tries to decrypt seq=1 without first receiving seq=0.
	if _, err := serverCh.Decrypt(ct1); err == nil {
		t.Fatal("strict mode should reject a gap (missing seq 0)")
	}
}

func TestNewDatagramChannel(t *testing.T) {
	key := make([]byte, 32)
	if _, err := rand.Read(key); err != nil {
		t.Fatal(err)
	}
	sendKey, err := hkdfDerive(key, []byte("s2c"))
	if err != nil {
		t.Fatal(err)
	}
	recvKey, err := hkdfDerive(key, []byte("c2s"))
	if err != nil {
		t.Fatal(err)
	}

	ch, err := NewDatagramChannel(sendKey, recvKey)
	if err != nil {
		t.Fatal(err)
	}
	if ch.mode != ModeDatagrams {
		t.Fatal("NewDatagramChannel should create a channel in ModeDatagrams")
	}
}

func TestDatagramModeSimulatedPacketLoss(t *testing.T) {
	clientCh, serverCh := newTestDatagramChannel(t)

	const total = 1000
	lossRates := []float64{0.05, 0.10, 0.20, 0.50}

	for _, lossRate := range lossRates {
		t.Run(fmt.Sprintf("loss_%.0f%%", lossRate*100), func(t *testing.T) {
			// Reset channels for each sub-test.
			clientCh, serverCh = newTestDatagramChannel(t)

			// Encrypt all messages.
			ciphertexts := make([][]byte, total)
			for i := range total {
				ciphertexts[i] = clientCh.Encrypt([]byte(fmt.Sprintf("msg-%d", i)))
			}

			// Simulate loss: drop packets deterministically based on index.
			// Use a simple hash to distribute drops evenly.
			delivered := 0
			dropped := 0
			var lastDelivered int
			for i, ct := range ciphertexts {
				// Deterministic "random" drop based on index.
				drop := float64((i*7+13)%100) / 100.0 < lossRate
				if drop {
					dropped++
					continue
				}

				plaintext, err := serverCh.Decrypt(ct)
				if err != nil {
					t.Fatalf("seq %d: decrypt failed after %d delivered, %d dropped: %v",
						i, delivered, dropped, err)
				}

				expected := fmt.Sprintf("msg-%d", i)
				if string(plaintext) != expected {
					t.Fatalf("seq %d: got %q, want %q", i, plaintext, expected)
				}
				delivered++
				lastDelivered = i
			}

			t.Logf("%.0f%% loss: %d/%d delivered, %d dropped, last seq=%d",
				lossRate*100, delivered, total, dropped, lastDelivered)

			if delivered == 0 {
				t.Fatal("no messages delivered")
			}
			if delivered+dropped != total {
				t.Fatalf("delivered(%d) + dropped(%d) != total(%d)", delivered, dropped, total)
			}
		})
	}
}

func TestDatagramModeStress(t *testing.T) {
	clientCh, serverCh := newTestDatagramChannel(t)

	// Rapid-fire 10,000 messages with random gaps.
	const total = 10000
	ciphertexts := make([][]byte, total)
	for i := range total {
		payload := make([]byte, 64)
		rand.Read(payload)
		ciphertexts[i] = clientCh.Encrypt(payload)
	}

	// Deliver only even-numbered packets (50% loss, every other packet).
	delivered := 0
	for i := 0; i < total; i += 2 {
		if _, err := serverCh.Decrypt(ciphertexts[i]); err != nil {
			t.Fatalf("seq %d: %v (after %d delivered)", i, err, delivered)
		}
		delivered++
	}

	t.Logf("stress: %d/%d delivered (50%% systematic loss)", delivered, total)
	if delivered != total/2 {
		t.Fatalf("expected %d delivered, got %d", total/2, delivered)
	}
}

func TestConfirmationCodeCrossplatformVector(t *testing.T) {
	// Fixed X25519 public keys (any 32-byte value is a valid X25519 point).
	keyA := bytes.Repeat([]byte{0x01}, 32)
	keyB := bytes.Repeat([]byte{0x02}, 32)

	pubA, err := ecdh.X25519().NewPublicKey(keyA)
	if err != nil {
		t.Fatal(err)
	}
	pubB, err := ecdh.X25519().NewPublicKey(keyB)
	if err != nil {
		t.Fatal(err)
	}

	code, err := DeriveConfirmationCode(pubA, pubB)
	if err != nil {
		t.Fatal(err)
	}

	// This expected value must match the TypeScript test in web/src/crypto.test.ts.
	const expected = "629624"
	if code != expected {
		t.Fatalf("got %q, want %q", code, expected)
	}

	// Verify order-independence with the same expected value.
	codeBA, err := DeriveConfirmationCode(pubB, pubA)
	if err != nil {
		t.Fatal(err)
	}
	if codeBA != expected {
		t.Fatalf("reversed order: got %q, want %q", codeBA, expected)
	}
}

func TestPairingRecordRoundTrip(t *testing.T) {
	// Simulate a pairing: two key pairs, record one side.
	serverKP, err := GenerateKeyPair()
	if err != nil {
		t.Fatal(err)
	}
	clientKP, err := GenerateKeyPair()
	if err != nil {
		t.Fatal(err)
	}

	record := NewPairingRecord("server-uuid", "https://relay.example.com", clientKP, serverKP.Public)

	// Marshal/unmarshal.
	data, err := record.Marshal()
	if err != nil {
		t.Fatal(err)
	}
	restored, err := UnmarshalPairingRecord(data)
	if err != nil {
		t.Fatal(err)
	}

	// Derive channel from restored record.
	ch, err := restored.DeriveChannel([]byte("c2s"), []byte("s2c"))
	if err != nil {
		t.Fatal(err)
	}

	// Derive the same channel from the original keys (server side).
	serverSendKey, err := DeriveSessionKey(serverKP.Private, clientKP.Public, []byte("s2c"))
	if err != nil {
		t.Fatal(err)
	}
	serverRecvKey, err := DeriveSessionKey(serverKP.Private, clientKP.Public, []byte("c2s"))
	if err != nil {
		t.Fatal(err)
	}
	serverCh, err := NewChannel(serverSendKey, serverRecvKey)
	if err != nil {
		t.Fatal(err)
	}

	// Verify they can communicate.
	ct := ch.Encrypt([]byte("hello from restored record"))
	pt, err := serverCh.Decrypt(ct)
	if err != nil {
		t.Fatal(err)
	}
	if string(pt) != "hello from restored record" {
		t.Fatalf("got %q", pt)
	}
}

func TestDeriveConfirmationCode(t *testing.T) {
	client, err := GenerateKeyPair()
	if err != nil {
		t.Fatal(err)
	}
	server, err := GenerateKeyPair()
	if err != nil {
		t.Fatal(err)
	}

	// Both sides compute the same code.
	code1, err := DeriveConfirmationCode(server.Public, client.Public)
	if err != nil {
		t.Fatal(err)
	}
	code2, err := DeriveConfirmationCode(client.Public, server.Public)
	if err != nil {
		t.Fatal(err)
	}
	if code1 != code2 {
		t.Fatalf("codes should be order-independent: %s vs %s", code1, code2)
	}
	if len(code1) != 6 {
		t.Fatalf("expected 6-digit code, got %q", code1)
	}

	// MitM: adversary substitutes its own key — codes diverge.
	adversary, err := GenerateKeyPair()
	if err != nil {
		t.Fatal(err)
	}
	codeServer, err := DeriveConfirmationCode(server.Public, adversary.Public)
	if err != nil {
		t.Fatal(err)
	}
	codeClient, err := DeriveConfirmationCode(adversary.Public, client.Public)
	if err != nil {
		t.Fatal(err)
	}
	if codeServer == codeClient {
		t.Fatal("MitM codes should differ (collision — extremely unlikely)")
	}
	if codeServer == code1 {
		t.Fatal("MitM server code should differ from honest code")
	}
	if codeClient == code1 {
		t.Fatal("MitM client code should differ from honest code")
	}
}

File: web/src/crypto.test.ts
// Copyright 2026 Marcelo Cantos
// SPDX-License-Identifier: Apache-2.0

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  E2EKeyPair,
  E2EChannel,
  deriveKeyFromSecret,
  deriveConfirmationCode,
} from "./crypto.js";

describe("E2EKeyPair", () => {
  it("generates a 32-byte public key", async () => {
    const kp = await E2EKeyPair.create();
    assert.equal(kp.publicKeyData.length, 32);
  });

  it("two key pairs derive the same session key", async () => {
    const alice = await E2EKeyPair.create();
    const bob = await E2EKeyPair.create();
    const info = new TextEncoder().encode("test-session");

    const keyA = await alice.deriveSessionKey(bob.publicKeyData, info);
    const keyB = await bob.deriveSessionKey(alice.publicKeyData, info);

    assert.deepEqual(keyA, keyB);
    assert.equal(keyA.length, 32);
  });
});

describe("E2EChannel", () => {
  it("encrypt/decrypt round trip", async () => {
    const key = globalThis.crypto.getRandomValues(new Uint8Array(32));
    const ch = await E2EChannel.fromSharedKey(key, false);
    // Create a matching channel for the other side
    const ch2 = await E2EChannel.fromSharedKey(key, true);

    const plaintext = new TextEncoder().encode("hello, tern!");
    const encrypted = await ch.encrypt(plaintext);
    const decrypted = await ch2.decrypt(encrypted);

    assert.deepEqual(decrypted, plaintext);
  });

  it("bidirectional channel communication", async () => {
    const key = globalThis.crypto.getRandomValues(new Uint8Array(32));
    const client = await E2EChannel.fromSharedKey(key, false);
    const server = await E2EChannel.fromSharedKey(key, true);

    // Client -> Server
    const msg1 = new TextEncoder().encode("from client");
    const enc1 = await client.encrypt(msg1);
    const dec1 = await server.decrypt(enc1);
    assert.deepEqual(dec1, msg1);

    // Server -> Client
    const msg2 = new TextEncoder().encode("from server");
    const enc2 = await server.encrypt(msg2);
    const dec2 = await client.decrypt(enc2);
    assert.deepEqual(dec2, msg2);

    // Multiple messages in same direction
    const msg3 = new TextEncoder().encode("second from client");
    const enc3 = await client.encrypt(msg3);
    const dec3 = await server.decrypt(enc3);
    assert.deepEqual(dec3, msg3);
  });

  it("rejects replay attack (repeated ciphertext)", async () => {
    const key = globalThis.crypto.getRandomValues(new Uint8Array(32));
    const client = await E2EChannel.fromSharedKey(key, false);
    const server = await E2EChannel.fromSharedKey(key, true);

    const plaintext = new TextEncoder().encode("test");
    const encrypted = await client.encrypt(plaintext);

    // First decrypt succeeds
    await server.decrypt(encrypted);

    // Replay should fail (sequence number already consumed)
    await assert.rejects(() => server.decrypt(encrypted), {
      message: "unexpected sequence number",
    });
  });

  it("rejects ciphertext too short", async () => {
    const key = globalThis.crypto.getRandomValues(new Uint8Array(32));
    const ch = await E2EChannel.fromSharedKey(key, true);

    await assert.rejects(() => ch.decrypt(new Uint8Array(4)), {
      message: "ciphertext too short",
    });
  });

  it("datagram mode accepts messages with gaps", async () => {
    const key = globalThis.crypto.getRandomValues(new Uint8Array(32));
    const client = await E2EChannel.fromSharedKey(key, false);
    const server = await E2EChannel.fromSharedKey(key, true);
    server.mode = "datagrams";

    // Encrypt seq 0..5 on sender.
    const cts: Uint8Array[] = [];
    for (let i = 0; i < 6; i++) {
      cts.push(await client.encrypt(new TextEncoder().encode(`msg${i}`)));
    }

    // Receiver gets seq 0, 1, 5 (skipping 2, 3, 4).
    for (const idx of [0, 1, 5]) {
      const pt = await server.decrypt(cts[idx]);
      assert.deepEqual(pt, new TextEncoder().encode(`msg${idx}`));
    }
  });

  it("datagram mode rejects replay", async () => {
    const key = globalThis.crypto.getRandomValues(new Uint8Array(32));
    const client = await E2EChannel.fromSharedKey(key, false);
    const server = await E2EChannel.fromSharedKey(key, true);
    server.mode = "datagrams";

    const ct = await client.encrypt(new TextEncoder().encode("hello"));
    await server.decrypt(ct);

    // Replay same ciphertext — must fail.
    await assert.rejects(() => server.decrypt(ct), {
      message: "sequence number replayed or too old",
    });
  });

  it("datagram mode rejects old sequence numbers", async () => {
    const key = globalThis.crypto.getRandomValues(new Uint8Array(32));
    const client = await E2EChannel.fromSharedKey(key, false);
    const server = await E2EChannel.fromSharedKey(key, true);
    server.mode = "datagrams";

    // Encrypt seq 0..5.
    const cts: Uint8Array[] = [];
    for (let i = 0; i < 6; i++) {
      cts.push(await client.encrypt(new TextEncoder().encode(`msg${i}`)));
    }

    // Receive seq=5 first (jump forward).
    await server.decrypt(cts[5]);

    // seq=3 is now in the past — must be rejected.
    await assert.rejects(() => server.decrypt(cts[3]), {
      message: "sequence number replayed or too old",
    });
  });

  it("strict mode rejects gaps", async () => {
    const key = globalThis.crypto.getRandomValues(new Uint8Array(32));
    const client = await E2EChannel.fromSharedKey(key, false);
    const server = await E2EChannel.fromSharedKey(key, true);

    // Encrypt two packets but skip delivering seq=0.
    await client.encrypt(new TextEncoder().encode("first"));
    const ct1 = await client.encrypt(new TextEncoder().encode("second"));

    // seq=1 without seq=0 must fail in strict mode.
    await assert.rejects(() => server.decrypt(ct1), {
      message: "unexpected sequence number",
    });
  });

  it("direct key construction", async () => {
    const sendKey = globalThis.crypto.getRandomValues(new Uint8Array(32));
    const recvKey = globalThis.crypto.getRandomValues(new Uint8Array(32));

    const sender = await E2EChannel.create(sendKey, recvKey);
    const receiver = await E2EChannel.create(recvKey, sendKey);

    const msg = new TextEncoder().encode("direct keys");
    const enc = await sender.encrypt(msg);
    const dec = await receiver.decrypt(enc);
    assert.deepEqual(dec, msg);
  });
});

describe("deriveKeyFromSecret", () => {
  it("derives a 32-byte key", async () => {
    const secret = globalThis.crypto.getRandomValues(new Uint8Array(32));
    const info = new TextEncoder().encode("test-info");
    const key = await deriveKeyFromSecret(secret, info);
    assert.equal(key.length, 32);
  });

  it("same inputs produce same output", async () => {
    const secret = new Uint8Array(32).fill(0x42);
    const info = new TextEncoder().encode("deterministic");
    const k1 = await deriveKeyFromSecret(secret, info);
    const k2 = await deriveKeyFromSecret(secret, info);
    assert.deepEqual(k1, k2);
  });

  it("different info produces different output", async () => {
    const secret = new Uint8Array(32).fill(0x42);
    const k1 = await deriveKeyFromSecret(
      secret,
      new TextEncoder().encode("info-a"),
    );
    const k2 = await deriveKeyFromSecret(
      secret,
      new TextEncoder().encode("info-b"),
    );
    assert.notDeepEqual(k1, k2);
  });
});

describe("deriveConfirmationCode", () => {
  it("produces a 6-digit string", async () => {
    const alice = await E2EKeyPair.create();
    const bob = await E2EKeyPair.create();
    const code = await deriveConfirmationCode(
      alice.publicKeyData,
      bob.publicKeyData,
    );
    assert.match(code, /^\d{6}$/);
  });

  it("is order-independent", async () => {
    const alice = await E2EKeyPair.create();
    const bob = await E2EKeyPair.create();
    const code1 = await deriveConfirmationCode(
      alice.publicKeyData,
      bob.publicKeyData,
    );
    const code2 = await deriveConfirmationCode(
      bob.publicKeyData,
      alice.publicKeyData,
    );
    assert.equal(code1, code2);
  });

  it("different key pairs produce different codes (probabilistic)", async () => {
    const a = await E2EKeyPair.create();
    const b = await E2EKeyPair.create();
    const c = await E2EKeyPair.create();
    const code1 = await deriveConfirmationCode(a.publicKeyData, b.publicKeyData);
    const code2 = await deriveConfirmationCode(a.publicKeyData, c.publicKeyData);
    // With 1M possible codes, collision is ~0.0001% — safe to assert inequality
    assert.notEqual(code1, code2);
  });
});

describe("full ECDH pairing flow", () => {
  it("completes key exchange and establishes encrypted channel", async () => {
    // 1. Both sides generate key pairs
    const alice = await E2EKeyPair.create();
    const bob = await E2EKeyPair.create();

    // 2. Exchange public keys and derive session keys
    const info = new TextEncoder().encode("pairing-session");
    const aliceKey = await alice.deriveSessionKey(bob.publicKeyData, info);
    const bobKey = await bob.deriveSessionKey(alice.publicKeyData, info);
    assert.deepEqual(aliceKey, bobKey);

    // 3. Derive confirmation codes — both should match
    const codeA = await deriveConfirmationCode(
      alice.publicKeyData,
      bob.publicKeyData,
    );
    const codeB = await deriveConfirmationCode(
      bob.publicKeyData,
      alice.publicKeyData,
    );
    assert.equal(codeA, codeB);

    // 4. Create symmetric channels
    const aliceCh = await E2EChannel.fromSharedKey(aliceKey, false);
    const bobCh = await E2EChannel.fromSharedKey(bobKey, true);

    // 5. Bidirectional communication
    const msg1 = new TextEncoder().encode("Hello from Alice");
    const enc1 = await aliceCh.encrypt(msg1);
    const dec1 = await bobCh.decrypt(enc1);
    assert.deepEqual(dec1, msg1);

    const msg2 = new TextEncoder().encode("Hello from Bob");
    const enc2 = await bobCh.encrypt(msg2);
    const dec2 = await aliceCh.decrypt(enc2);
    assert.deepEqual(dec2, msg2);
  });
});
